[lld][MachO] rename to bp-* options for SectionOrderer

Colibrow opened the pull request:
Rename options related to profile guided function order (#96268) to prepare for the addition to the ELF port.

ellishg commented on lld/test/ELF/bp-section-orderer-err.s:
```diff
@@ -0,0 +1,44 @@
+# REQUIRES: aarch64
```
See my comment here
https://github.com/llvm/llvm-project/pull/117514/files#r1866406570

Colibrow commented:
@ellishg hi ellis, I've fixed these. could you please review again?

ellishg commented on lld/MachO/Driver.cpp:
```diff
@@ -1838,34 +1838,44 @@ bool link(ArrayRef<const char *> argsArr, llvm::raw_ostream &stdoutOS,
-    config->irpgoProfileSortProfilePath = arg->getValue();
+  if (args.hasArg(OPT_irpgo_profile_sort) ||
+      args.hasArg(OPT_irpgo_profile_sort_eq))
+    error("--irpgo-profile-sort is deprecated. Please use "
+          "--bp-startup-sort=function");
```
Let's make this a warning so the flag is still supported. Also, these flag still need to set `config->irpgoProfilePath` and `config->bpStartupFunctionSort` so they still work.
```suggestion
    warn("--irpgo-profile-sort is deprecated. Please use "
          "--bp-startup-sort=function");
```

Colibrow replied:
@ellishg Because the lld testcase add the `-fatal_warnings` so if I change to the `log()`. Is it OK?

thevinster replied:
You can use `%no-fatal-warnings-lld` instead `%lld` for the test case if you want to use lld without `-fatal_warnings`.

ellishg commented on lld/MachO/Options.td:
```diff
@@ -126,17 +126,21 @@ def no_call_graph_profile_sort : Flag<["--"], "no-call-graph-profile-sort">,
+    Group<grp_lld>;
 def irpgo_profile_sort: Separate<["--"], "irpgo-profile-sort">, Group<grp_lld>;
 def irpgo_profile_sort_eq: Joined<["--"], "irpgo-profile-sort=">,
-    Alias<!cast<Separate>(irpgo_profile_sort)>, MetaVarName<"<profile>">,
```
Let's keep this alias for now.

ellishg commented on lld/MachO/Driver.cpp:
```diff
@@ -1838,34 +1838,44 @@ bool link(ArrayRef<const char *> argsArr, llvm::raw_ostream &stdoutOS,
+          "--bp-startup-sort=function");
+  if (const Arg *arg = args.getLastArg(OPT_irpgo_profile_eq)) {
+    config->irpgoProfilePath = arg->getValue();
     IncompatWithCGSort(arg->getSpelling());
```
I'm open to discussion, but I think `-irpgo-profile=` could be compatible with CGSort. It's only if we also have `-bp-start-sort=function` where this is a problem, which is checked later.
```suggestion
```

ellishg commented on lld/MachO/Options.td:
```diff
@@ -126,17 +126,21 @@ def no_call_graph_profile_sort : Flag<["--"], "no-call-graph-profile-sort">,
+    HelpText<"Deprecated. Please use --bp-startup-sort=function">, Group<grp_lld>;
 def compression_sort_startup_functions: Flag<["--"], "compression-sort-startup-functions">,
-    HelpText<"Order startup functions to improve compressed size in addition to startup time">,
-    Group<grp_lld>;
```
`-compression-sort-startup-functions` is a separate flag from `-bp-startup-sort=function`. This is saying that for the functions that we are ordering for startup, we also want to order them for compression. `-bp-startup-sort` is saying we want to order functions for startup. We should leave `compression_sort_startup_functions` alone for now, although I am open to different spellings, especially since we might order data for startup in the future.

ellishg commented on lld/MachO/Options.td:
```diff
@@ -126,17 +126,21 @@ def no_call_graph_profile_sort : Flag<["--"], "no-call-graph-profile-sort">,
+  Alias<bp_startup_sort>, AliasArgs<["function"]>, HelpText<"Deprecated. Please use --bp_startup_sort=function">, Group<grp_lld>;
 def no_compression_sort_startup_functions: Flag<["--"], "no-compression-sort-startup-functions">,
-    HelpText<"Do not order startup function for compression">, Group<grp_lld>;
-def compression_sort: Joined<["--"], "compression-sort=">,
```
We should keep this flag around and alias it to `-bp-compression-sort`. The point of keeping these flags is that I want the old linker flags to keep working after this change. This PR should not break any prior behavior.

Colibrow replied:
got it.

ellishg commented on lld/test/ELF/incompatible.s:
```diff
@@ -1,3 +1,16 @@
+# RUN: rm -rf %t
+# RUN: touch %t.call_graph 
+# RUN: touch %t.profdata 
+# RUN: not ld.lld -o /dev/null %t --irpgo-profile=%t.profdata --call-graph-ordering-file=%t.call_graph 2>&1 | FileCheck %s --check-prefix=IRPGO-ERR
```
```suggestion
# RUN: not ld.lld -o /dev/null %t --irpgo-profile=%t.profdata --bp-startup-sort=function --call-graph-ordering-file=%t.call_graph 2>&1 | FileCheck %s --check-prefix=IRPGO-ERR
```

ellishg commented on lld/test/MachO/bp-section-orderer-errs.s:
```diff
@@ -1,12 +1,15 @@
-# RUN: not %lld -o /dev/null --irpgo-profile-sort %s --call-graph-profile-sort 2>&1 | FileCheck %s --check-prefix=IRPGO-ERR
-# RUN: not %lld -o /dev/null --irpgo-profile-sort=%s --call-graph-profile-sort 2>&1 | FileCheck %s --check-prefix=IRPGO-ERR
-# IRPGO-ERR: --irpgo-profile-sort is incompatible with --call-graph-profile-sort
+# RUN: not %lld -o /dev/null --irpgo-profile=%s --call-graph-profile-sort 2>&1 | FileCheck %s --check-prefix=IRPGO-ERR
```
```suggestion
# RUN: not %lld -o /dev/null --irpgo-profile=%s -bp-startup-sort=function --call-graph-profile-sort 2>&1 | FileCheck %s --check-prefix=IRPGO-ERR
```

ellishg commented on lld/test/ELF/incompatible.s:
```diff
@@ -1,3 +1,16 @@
+# REQUIRES: aarch64
+# RUN: rm -rf %t
+# RUN: touch %t.call_graph 
+# RUN: touch %t.profdata 
```
Minor nit, but I realized that you can use `/dev/null` in place of these files and remove the `touch` commands.

ellishg commented on lld/test/MachO/bp-section-orderer.s:
```diff
@@ -4,12 +4,12 @@
 
-# RUN: %lld -arch arm64 -lSystem -e _main -o %t/a.out %t/a.o --irpgo-profile-sort=%t/a.profdata --verbose-bp-section-orderer 2>&1 | FileCheck %s --check-prefix=STARTUP
-# RUN: %lld -arch arm64 -lSystem -e _main -o %t/a.out %t/a.o --irpgo-profile-sort=%t/a.profdata --verbose-bp-section-orderer --icf=all --compression-sort=none 2>&1 | FileCheck %s --check-prefix=STARTUP
+# RUN: %lld -arch arm64 -lSystem -e _main -o %t/a.out %t/a.o --irpgo-profile=%t/a.profdata --verbose-bp-section-orderer 2>&1 | FileCheck %s --check-prefix=STARTUP
```
The `-bp-startup-sort=function` flag is required to tell BP to order functions while `--irpgo-profile` is simply used to provide the profile. Same for lines 8 and 12.

```suggestion
# RUN: %lld -arch arm64 -lSystem -e _main -o %t/a.out %t/a.o --irpgo-profile=%t/a.profdata -bp-startup-sort=function --verbose-bp-section-orderer 2>&1 | FileCheck %s --check-prefix=STARTUP
```

ellishg replied:
To make sure the prior flags still work, I think we should just copy these RUN lines and leave the original ones as they are.

MaskRay commented:
The ELF change should be reverted. The purpose of the PR is to add some bp- options and make existing options (deployed by Meta) aliases (for compatibility). The ELF port should not be touched as it doesn't have the functionality yet.

Colibrow committed:
[lld][MachO] rename to bp-* options for SectionOrderer in MachO

Colibrow commented:
@MaskRay @ellishg Hi, I've removed the ELF parts(which will be added in another PR where it's been used in fact). And fixed the problems. Could you please review this PR?

ellishg reviewed (commented):
Thanks for addressing all our comments. I'm starting to like this set of flags a lot more than before

Colibrow commented on lld/MachO/SectionPriorities.cpp:
```diff
@@ -353,14 +353,16 @@ void macho::PriorityBuilder::parseOrderFile(StringRef path) {
   DenseMap<const InputSection *, size_t> sectionPriorities;
-  if (!config->irpgoProfileSortProfilePath.empty() ||
-      config->functionOrderForCompression || config->dataOrderForCompression) {
+  if (!config->irpgoProfilePath.empty() ||
```
is it `if (config->bpStartupFunctionSort`

ellishg replied:
Oh yes, sorry

ellishg reviewed (approved):
LGTM

Colibrow commented:
@MaskRay hi, can you please review this?

thevinster reviewed (approved):
Thanks!

Colibrow commented:
@MaskRay @ellishg Hi, Can you merge this PR? I don't have the write access.

MaskRay commented:
Note: the patch description is used as the default commit message, so please make sure to update it after the patch has been rewritten. I've changed the description.
